### ISSUE TITLE
feature request: allow sharing templates, resources, scripts, helpers between jsreport and toner

### ISSUE THREAD

**Issue by bjrmatos** [jsreport/jsreport#103]
Hi, i have a use case where i need to share my templates, scripts, helpers, etc between jsreport and [jsreport/toner](https://github.com/jsreport/toner)

i want to package my templates, scripts, helpers, etc in some kind of module (installable by npm) and share it with other applications, for example i want to use a template created with jsreport studio inside a [electron](http://electron.atom.io/) based desktop application, but i don't want/need the full jsreport as a dependency of my app, i just need jsreport/toner. 

Because jsreport saves templates, etc inside a db (nedb, mongodb, etc) i won't be able to share my templates, etc between projects, i know that i can read this information from db (nedb, mongodb, etc) and manually import them in my jsreport/toner configuration, but this won't work if any of this apps don't have access to the jsreport db.

So basically what i'm asking is to support sharing this kind of resources (templates, scripts, file resources) in a modular/node.js way, in order to do this i think jsreport could import this resources from my shared modules and do its job, right know the only problem i notice is how to sync the edit of a file inside jsreport studio with my module source's code.

if i understand correctly, right now template helpers are declared inside template definition, is there a way to define templates globally and let jsreport import them? so we can share template helpers between jsreport and jsreport/toner.

i think first we must identify what kind of information can be shared between jsreport and jsreport/toner, allowing share templates, scripts, helpers, import them , etc will be very useful IMHO.

**Comment by pofider** [comment]
We could write a jsreport extension which would react on jsreport data changes and transform these changes into node/npm/toner compatible js files. 

Or the toner application could request templates through HTTP requests from jsreport server.

Just note that toner not yet support resources or custom scripts.

**Comment by bjrmatos** [comment]
i'm interesting in a locally solution, so react on jsreport data changes seems the way to go.. i'll dive into it, thnks for your feedback :)

**Comment by pofider** [comment]
This could help...

```js
var templates = reporter.documentStore.collections["templates"];

templates.beforeInsertListeners.add("syncing", this, function(doc) {
  //write the doc to file
});                 

templates.beforeUpdateListeners.add("syncing", this , function(query, options) {
  //we would rather need after update, but you could do setTimeout for a second and then query the whole template and update your js file
});           
```

**Comment by bjrmatos** [comment]
awesome, let's see if i can finally contribute jahja

**Comment by bjrmatos** [comment]
do you have a comment for this?



> if i understand correctly, right now template helpers are declared inside template definition, is there a way to define templates globally and let jsreport import them?

**Comment by pofider** [comment]
Support for global helpers is quite required jsreport feature. Right now it can be achieved with a custom extension globally or by attaching a custom script doing this to every template.

The best would be probably to add a flag "isGlobal" to custom script entity. Then the custom script can run for every template if specified and add the helpers or do various other tasks.

**Comment by bjrmatos** [comment]
closing, because `jsreport-core` covers this
